### DISCUSSION TITLE
compile linkw and libw

### DISCUSSION THREAD

**Discussion by jklwn4** [nidud/asmc#14] [General]
Hi nidud,

how would i (re-)compile "linkw" and "libw" from the sources you supply - seems i need open watcom? I would prefer VS, any chance to do that?

JK

**Comment by nidud** [comment]
> how would i (re-)compile "linkw" and "libw" from the sources you supply

If you install https://github.com/open-watcom as `\watcom` you should be good to go.

> I would prefer VS, any chance to do that?

Well, you could use LIB and LINK I guess; LIBW and LINKW comes with some hassle.

There have however been a lot of [changes](https://github.com/nidud/asmc/blob/master/doc/command/option-link.md) there lately.

**Comment by jklwn4** [comment]
Ok - i installed watcom to D:\watcom. There is no project file (.wpj), so i should make use of the associated "makefile". I found "wmake" in binnt64, so i tried "wmake ...\makefile" inside a command prompt - to no avail. I get "Target (D:\linkw\master\libw\makefile) not mentioned in any makefile" and i have next to no experience using make utilities.

What am i doing wrong?

**Reply by nidud** [reply]
> Ok - i installed watcom to D:\watcom. There is no project file (.wpj), so i should make use of the associated "makefile". I found "wmake" in binnt64, so i tried "wmake ...\makefile" inside a command prompt - to no avail. I get "Target (D:\linkw\master\libw\makefile) not mentioned in any makefile" and i have next to no experience using make utilities.
> 
> What am i doing wrong?

There's a make utility included in Asmc. Just browse to the directory, select the makefile and hit enter. You need to build the `./lib` first as this is used by `./libw` and `./linkw`.

**Comment by nidud** [comment]
Note that there is a [/NODEC](https://github.com/nidud/asmc/blob/master/doc/tools/lib.md) option added to LIBW for the .DEF input.
```
LIB[W] /MACHINE:X86 /DEF:def_file /OUT:lib_file

```
The input is half-decorated (symbol@4) where an underscore is added (_symbol@4).
```
LINK[W] /DLL /MACHINE:X86 /DEF:def_file /OUT:dll_file *.obj
```
Here the .DEF input is non-decorated names (symbol). The linker will use the decoration from the object files to produce the import library for the DLL. LINKW (which spawns LIBW) use the option /NODEC for this purpose.

**Reply by jklwn4** [reply]
Thanks, i´m going to try that - maybe there will be more questions ...

**Comment by jklwn4** [comment]
Alas, i keep getting this: 

> D:\ASMC\bin>make D:\linkw\master\lib\makefile
Asmc Program Maintenance Utility Version 2.1.2.34
makefile: No such file or directory

Please, could you give more and more precise instructions how to do it, seems i don´t get it.

**Reply by nidud** [reply]
> Alas, i keep getting this:
> 
> > D:\ASMC\bin>make D:\linkw\master\lib\makefile
> > Asmc Program Maintenance Utility Version 2.1.2.34
> > makefile: No such file or directory

> Just browse to the directory, select the makefile and hit enter.

You need to be in the LIB directory.
No arguments to make: it uses the makefile in the current directory.

**Comment by jklwn4** [comment]
Ok, i browse to the LIB directory (D:\linkw\master\lib), i can see two items in Explorer: "makefile" and "src" (which is the folder containing the sources), now i select makefile and hit "enter"

I get a well known dialog ("how should this file be opened"), i browse to D:\ASMC\bin and choose "make.exe" - the screen flickers a bit - nothing else happens.

I think this requires some kind of setup like setting environment variables, running a batch beforehand or something like that ...
Where should a result of this be found at? How should make.exe know where \watcom is located? It needs watcom for compiling and linking, this is what you told me. Where should i place the files in linkw-master.zip to be found by make.exe? 

Sorry for being an absolute noob, i never have used watcom or (watcom/asmc) makefiles before, all of this is an absolute mistery to me

**Reply by nidud** [reply]
> Ok, i browse to the LIB directory (D:\linkw\master\lib), i can see two items in Explorer: 

Right-click your desktop, select New->Shortcut, browse to the D:\ASMC directory, select the asmc-2.35.cmd file (or bin\dz.exe). Click Ok. Run the shortcut.

**Reply by jklwn4** [reply]
Aha, i should have made use of Doszip Commander (!). So i browse to the makefile (D:\linkw\master\lib\makefile), select the makefile and hit enter. This is what i get:

> Asmc Program Maintenance Utility Version 2.1.2.34
        \WATCOM\binnt\wcc386 -I\WATCOM\h;src\h -q -zc -bc -bt=nt -ox -s -DINSIDE_WLINK -DNDEBUG -DIDE_PGM src\dwlngen.c
Der Befehl "\WATCOM\binnt\wcc386" ist entweder falsch geschrieben oder
konnte nicht gefunden werden.
src\dwlngen.obj: No error
(the command \WATCOM\binnt\wcc386" is either wrong or could not be found)

wcc386 is in \binnt64 in my installation, i have a 64 bit windows, so i installed the 64 bit version of Watcom after i got a hint to do so when first trying to install the 32 bit version - maybe this is the problem. Should i install the 32 bit version instead ?

We are getting closer ...

**Reply by nidud** [reply]
> > (the command \WATCOM\binnt\wcc386" is either wrong or could not be found)
> wcc386 is in \binnt64 in my installation, i have a 64 bit windows, so i installed the 64 bit version of Watcom after i got a hint to do so when first trying to install the 32 bit version - maybe this is the problem. 

Strange, there should be both
```
\WATCOM\binnt\wcc386.exe:
 Machine:       I386
 ImageBase:     00400000

\WATCOM\binnt64\wcc386.exe:
 Machine:               AMD64
 ImageBase:             0000000140000000
```
> Should i install the 32 bit version instead ?

Probably, the source for linkw is 32 bit.

**Comment by jklwn4** [comment]
Thanks for your help, i finally got it working. My misconception was: i should have used Doszip Commander and i should have made a full installation of watcom
